community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/SpatialIndexCache.java

```diff
@@ -56,17 +58,51 @@
      */
     T uncheckedSelect( CoordinateReferenceSystem crs )
     {
-        return spatials.computeIfAbsent( crs, key ->
+        T existing = spatials.get( crs );
+        if ( existing != null )
         {
-            try
-            {
-                return factory.newSpatial( crs );
-            }
-            catch ( IOException e )
+            return existing;
+        }
+
+        // Instantiate from factory. Do this under lock so that we coordinate with any concurrent call to close.
+        // Concurrent calls to instantiating parts won't contend with each other since there's only
+        // a single writer at a time anyway.
+        acquireInstantiateCloseLock();
+        try
+        {
+            return spatials.computeIfAbsent( crs, key ->
             {
-                throw new UncheckedIOException( e );
-            }
-        } );
+                try
+                {
+                    return factory.newSpatial( crs );
+                }
+                catch ( IOException e )
+                {
+                    throw new UncheckedIOException( e );
+                }
+            } );
+        }
+        finally
+        {
+            instantiateCloseLock.unlock();
+        }
+    }
+
+    private void acquireInstantiateCloseLock()
+    {
+        instantiateCloseLock.lock();
+        if ( closed )
+        {
+            instantiateCloseLock.unlock();
+            throw new IllegalStateException( this + " is already closed" );
+        }
+    }
+
+    void shutInstantiateCloseLock()
```

Comment:
I felt confused about this name. Maybe something like `closeUnderInstantiateCloseLock`?

Reply:
Wanted to avoid close twice, but it was named like that initially. I can change to it.